README.md

```diff
@@ -611,48 +612,71 @@ We can also extend the `focusReset` option with other behaviors in the future. H
 - `focusReset: "immediate"`: immediately resets the focus to the `<body>` element, without waiting for the promise to settle.
 - `focusReset: "two-stage"`: immediately resets the focus to the `<body>` element, and then has the same behavior as `"after-transition"`.
 
+#### Scrolling to fragments and scroll resetting
+
+Current single-page app navigations leave the user's scroll position where it is. This is true even if you try to navigate to a fragment, e.g. by doing `history.pushState("/article#subheading")`. The latter has caused significant pain in client-side router libraries; see e.g. [remix-run/react-router#394](https://github.com/remix-run/react-router/issues/394), or the manual code that is needed to handle this case in [Vue](https://sourcegraph.com/github.com/vuejs/router/-/blob/src/scrollBehavior.ts?L81-140), [Angular](https://github.com/angular/angular/blob/main/packages/router/src/router_scroller.ts#L76-L77), [React Router Hash Link](https://github.com/rafgraph/react-router-hash-link/blob/main/src/HashLink.jsx), and others.
+
+With the navigation API, there is a different default behavior, controllable via another option to `navigateEvent.intercept()`:
+
+- `e.intercept({ handler, scroll: "after-transition" })`: the default behavior. After the promise returned by `handler` fulfills, the browser will attempt to scroll to the fragment given by `e.destination.url`, or if there is no fragment, it will reset the scroll position to the top of the page (like in a cross-document navigation).
+- `e.intercept({ handler, scroll: "manual" })`: the browser will not change the user's scroll position, although you can later perform the same logic manually using `e.scroll()`.
+
+The `navigateEvent.scroll()` method could be useful if you know you have loaded the element referenced by the hash, or if you know you want to reset the scroll position to the top of the document early, before the full transition has finished. For example:
+
+```js
+if (navigateEvent.navigationType === "push" || navigateEvent.navigationType === "replace") {
+  navigateEvent.intercept({
+    scroll: "manual",
+    async handler() {
+      await fetchDataAndSetUpDOM(navigateEvent.url);
+      navigateEvent.scroll();
+
+      // Note: navigateEvent.scroll() will update what :target points to.
+      await fadeInTheScrolledToElement(document.querySelector(":target"));
```

Comment:
> Isn't this something that should happen after handling the navigation, rather than as part of it?

I'm not sure. It seems like you might want the loading spinner and AT announcement delayed until after the content is fully faded? I dunno though, the fade here is pretty artificial.

> ```js
>     // Although .scroll() was called before the DOM was populated, the browser
>    // will continually attempt to scroll based on heuristics.
> ```

Hmm. This was not the behavior I was planning (for push/replace). In particular I was planning to just trigger "scroll to fragment", not "try to scroll to fragment".

I can see the attraction (it's more like MPAs!) of doing the continual trying approach. But the whole "try to scroll to fragment" algorithm is rather non-deterministic and I kinda prefer the predictability of just "scroll to fragment"...

Reply:
Is there a middle ground where the developer could call `.scroll` multiple times when they think things may have moved?

Reply:
Nate and I discussed this and think for now it's simplest to have scroll() be one-shot, and to disallow it being called multiple times. However we can open it up to be called more than once in the future, pretty easily.